### PR TITLE
[Issue 84]fix cpu and mem in docker for Linux

### PR DESCRIPTION
**Pull request by overstep123**
1. fix nameservice to get correct cpu and mem resource from cgroups.
2. fix the KEY url in the nameservice Dockerfile. 
3. add pod anti affinity for broker for better high availability.(for example, if the two master broker are both in one node and the node is down,then there is a few time that the rocket cannot handle producing.)

### CODE REVIEW

File: images/namesrv/alpine/Dockerfile [hunk 1]
```diff
@@ -34,7 +34,7 @@ RUN set -eux; \
     apk add --virtual .build-deps curl gnupg unzip; \
     curl https://archive.apache.org/dist/rocketmq/${ROCKETMQ_VERSION}/rocketmq-all-${ROCKETMQ_VERSION}-bin-release.zip -o rocketmq.zip; \
     curl https://archive.apache.org/dist/rocketmq/${ROCKETMQ_VERSION}/rocketmq-all-${ROCKETMQ_VERSION}-bin-release.zip.asc -o rocketmq.zip.asc; \
-	curl https://www.apache.org/dist/rocketmq/KEYS -o KEYS; \
+	curl https://downloads.apache.org/rocketmq/KEYS -o KEYS; \
```
Comment (caigy) [inline images/namesrv/alpine/Dockerfile:37]: Could you provide the reason to change it? It seems that this modification is not related to the title

File: pkg/controller/broker/broker_controller.go [hunk 2]
```diff
@@ -375,14 +375,43 @@ func getBrokerName(broker *rocketmqv1alpha1.Broker, brokerGroupIndex int) string
...
+			PreferredDuringSchedulingIgnoredDuringExecution: []corev1.WeightedPodAffinityTerm{corev1.WeightedPodAffinityTerm{
+				Weight: 100,
+				PodAffinityTerm: corev1.PodAffinityTerm{
+					TopologyKey: "kubernetes.io/hostname",
+					LabelSelector: &metav1.LabelSelector{
+						MatchLabels: labelsForBroker(broker.Name),
+					},
+				},
+			}},
+		}
```
Comment (caigy) [inline pkg/controller/broker/broker_controller.go:401]: Changes should be related to this PR. Also, podAntiAffinity feature is discussed in #38 with PR  #75, you can  join it and work together to implement this feature.

File: images/namesrv/alpine/runserver-customize.sh [hunk 3]
```diff
@@ -54,8 +54,8 @@ calculate_heap_sizes()
 {
     case "`uname`" in
         Linux)
-            system_memory_in_mb=`free -m| sed -n '2p' | awk '{print $2}'`
-            system_cpu_cores=`egrep -c 'processor([[:space:]]+):.*' /proc/cpuinfo`
+            system_memory_in_mb=$(($(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)/1024/1024))
+            system_cpu_cores=$(($(cat /sys/fs/cgroup/cpu/cpu.cfs_quota_us)/100000))
```
Comment (caigy) [inline images/namesrv/alpine/runserver-customize.sh:58]: It would be better to check memory and cpu from cgroup are smaller than system memory and cpu, because resource limits may not be set.

**Review by caigy — CHANGES_REQUESTED** [review]
Also link the PR with related issue and rename the title starting with '[ISSUE xx]'

File: images/namesrv/alpine/Dockerfile (hunk 1 above)
Comment (overstep123) [reply images/namesrv/alpine/Dockerfile:37]: this is the result of the old url:
image

File: images/namesrv/alpine/runserver-customize.sh (hunk 3 above)
Comment (overstep123) [reply images/namesrv/alpine/runserver-customize.sh]: This is better,and I fixed it.Although i think it's not correct if anone use it in the kubernetes without memory limit.

File: pkg/controller/broker/broker_controller.go (hunk 2 above)
Comment (overstep123) [reply pkg/controller/broker/broker_controller.go:401]: I  have no idea why the nodeAffinity is needed. But the podAntiAffinity is sure needed. Because I don't want to see that the whole cluster(maybe multi master borker) become unwritable when all the master broker are assigned to the same node and the node is down. This situation will be worse if the kubernetes can't new the master pod in time.

File: images/namesrv/alpine/Dockerfile (hunk 1 above)
Comment (caigy) [reply images/namesrv/alpine/Dockerfile:37]: Ok, this is indeed a problem. I think you can use `curl -L` to solve the redirect problem, which is the same as that in rocketmq-docker repo.

File: pkg/controller/broker/broker_controller.go (hunk 2 above)
Comment (caigy) [reply pkg/controller/broker/broker_controller.go:401]: You'b better open another PR to do it, because this is obviously unrelated to this issue.

File: images/namesrv/alpine/runserver-customize.sh [hunk 4]
```diff
@@ -55,7 +55,15 @@ calculate_heap_sizes()
     case "`uname`" in
         Linux)
             system_memory_in_mb=`free -m| sed -n '2p' | awk '{print $2}'`
+            system_memory_in_mb_in_docker=$(($(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)/1024/1024))
+            if [ $system_memory_in_mb_in_docker -lt $system_memory_in_mb ];then
+              system_memory_in_mb=$system_memory_in_mb_in_docker
+            fi
             system_cpu_cores=`egrep -c 'processor([[:space:]]+):.*' /proc/cpuinfo`
+            system_cpu_cores_in_docker=$(($(cat /sys/fs/cgroup/cpu/cpu.cfs_quota_us)/100000))
```
[the remark this answers is not in the document]
Comment (overstep123) [reply images/namesrv/alpine/runserver-customize.sh]: that's right

**Review by caigy — APPROVED** [review]

**Comment by caigy** [comment]
@overstep123 Pls link the PR to related issue and rename the title starting with '[ISSUE xx]'.

**Review by gobbq — APPROVED** [review]

**Review by gobbq — APPROVED** [review]

**Review by duhenglucky — APPROVED** [review]
